src/OneDriveSdk/Helpers/UploadSessionHelper.cs

```diff
@@ -0,0 +1,257 @@
+                    
+                    yield return uploadRequest;
+
+                    currentRangeBegins += nextChunkSize;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get the status of the session. Stores returned session internally.
+        /// Updates internal list of ranges remaining to be uploaded (according to the server).
+        /// </summary>
+        /// <returns>UploadSession returned by the server.</returns>
+        public async Task<UploadSession> UpdateSessionStatusAsync()
+        {
+            var request = new UploadSessionRequest(this.Session, this.client, null);
+            var newSession = await request.GetAsync();
+            
+            var newRangesRemaining = this.GetRangesRemaining(newSession);
+
+            this.rangesRemaining = newRangesRemaining;
+            newSession.UploadUrl = this.Session.UploadUrl; // Sometimes the UploadUrl is not returned
+            this.Session = newSession;
+            return newSession;
+        }
+
+        private List<Tuple<int, int>> GetRangesRemaining(UploadSession session)
+        {
+            // nextExpectedRanges: https://dev.onedrive.com/items/upload_large_files.htm
+            // Sample: ["12345-55232","77829-99375"]
+            // Also, second number in range can be blank, which means 'until the end'
+            var newRangesRemaining = new List<Tuple<int, int>>();
+            foreach (var range in session.NextExpectedRanges)
+            {
+                var rangeSpecifiers = range.Split('-');
+                newRangesRemaining.Add(new Tuple<int, int>(int.Parse(rangeSpecifiers[0]),
+                    string.IsNullOrEmpty(rangeSpecifiers[1]) ? this.totalUploadLength - 1 : int.Parse(rangeSpecifiers[1])));
+            }
+
+            return newRangesRemaining;
+        }
+
+        /// <summary>
+        /// Delete the session.
+        /// </summary>
+        /// <returns>Once returned task is complete, the session has been deleted.</returns>
+        public async Task DeleteSession()
+        {
+            var request = new UploadSessionRequest(this.Session, this.client, null);
+            await request.DeleteAsync();
+        }
+
+        /// <summary>
+        /// Upload the whole session.
+        /// </summary>
+        /// <param name="maxTries">Number of times to retry entire session before giving up.</param>
+        /// <returns>Item information returned by server.</returns>
+        public async Task<Item> UploadAsync(int maxTries = 3, IEnumerable<Option> options = null)
+        {
+            var uploadTries = 0;
+            var readBuffer = new byte[this.maxChunkSize];
+            
+            while (uploadTries < maxTries)
+            {
+                var chunkRequests = this.GetUploadChunkRequests(options);
+
+                foreach (var request in chunkRequests)
+                {
+                    var tries = 0;
+                    var requestSucceeded = false;
```

Comment:
you can compress this into a single bool, firstAttempt = true, if firstAttempt == false then you throw